### PR TITLE
feat: Added new goal to verify config file is present and readable.

### PR DESCRIPTION
**Pull request by marcuss**
Fixes ISSUE #5 - Validates the config file exists and is readable otherwise throw a configuration error and skip mojo execution.
Also 
Fixes ISSUE #13 - Perform appmap.yml validation

### CODE REVIEW

File: src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java [hunk 1]
```diff
@@ -5,28 +5,39 @@
...
-                skipMojo();
+                skipMojo("Skipping AppLand AppMap execution because property skip is set.");
                 return;
+            } else if (!isConfigFileValid()) {
+                skipMojo(
+                        "Skipping AppLand AppMap execution because the config file: "
+                                + configFile.getPath()
+                                + " does not exist or is not readable"
+                );
+                throw new FileNotFoundException(configFile.getPath());
```
Comment (dustinbyrne) [inline src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java:29]: Is it possible to roll this up into a separate goal that `prepare-agent` depends on? It'd be useful for troubleshooting.

File: build.gradle [hunk 2]
```diff
@@ -18,7 +18,7 @@ plugins {
 def defGroupId       = 'com.appland'
 def defArtifactId    = 'appmap-maven-plugin'
 def defName          = 'AppMap Maven Plugin'
-def defVersion       = "1.1.0"
+def defVersion       = "1.1.1"
```
Comment (dustinbyrne) [inline build.gradle:21]: Can you drop this change and the change to `CHANGELOG.md`? We're using automation to manage this now so you no longer need to do this :)

Comment (dustinbyrne) [reply build.gradle:21]: To be clear, the automation is responsible for bumping the version and adding CHANGELOG entries. See an example here: https://github.com/applandinc/appmap-maven-plugin/commit/34cb01a11daef70f6d620d0b9baab6ff8884a3ea

**Review by dustinbyrne — CHANGES_REQUESTED** [review]

File: src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java (hunk 1 above)
Comment (marcuss) [reply src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java:29]: Creating a goal to check the file is correct? something like ´´validate-config´´ will this be compatible with your issue about validating the configuration content?

File: build.gradle (hunk 2 above)
Comment (marcuss) [reply build.gradle:21]: done

File: src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java (hunk 1 above)
Comment (marcuss) [reply src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java:29]: The only thing missing is this one, what if we rather open another issue to it ? I'm trying to focus on gradle plugin and will get back to the maven plugin soon.

**Comment by marcuss** [comment]
@dustinbyrne 

Dustin, I added the suggested new goal for the plugin, so people can troubleshoot its set up, altho it is an independent goal.

Check this out when found:


` `mvn com.appland:appmap-maven-plugin:validate-config``
```

[INFO] -----------------< co.marcuss.hipcontacts:hipcontacts >-----------------
[INFO] Building hipcontacts 0.0.1-SNAPSHOT
[INFO] --------------------------------[ war ]---------------------------------
[INFO]
[INFO] --- appmap-maven-plugin:1.1.1:validate-config (default-cli) @ hipcontacts ---
[INFO] Appland AppMap Configuration file found.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.544 s
[INFO] Finished at: 2021-04-07T21:27:59-05:00
[INFO] ------------------------------------------------------------------------
```


When not found:

` `mvn com.appland:appmap-maven-plugin:validate-config``
```

[INFO] --- appmap-maven-plugin:1.1.1:validate-config (default-cli) @ hipcontacts ---
[INFO] Skipping AppLand AppMap execution because the config file: appmap.yml does not exist or is not readable
[ERROR] Error executing AppLand AppMap Java Recorder
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.519 s
[INFO] Finished at: 2021-04-07T21:29:24-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.appland:appmap-maven-plugin:1.1.1:validate-config (default-cli) on project hipcontacts: Error initializing AppLand AppMap Java Recorder: appmap.yml -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

File: src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java (hunk 1 above)
Comment (marcuss) [reply src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java:29]: this was solved, the goal  `validate-config` was added to Validate the config file is present and redable

**Comment by dustinbyrne** [comment]
Are we missing a dependency?

```sh
$ ./gradlew clean build

> Task :compileJava
Note: /Users/dustin/dev/appland/appmap-maven-plugin/src/main/java/com/appland/appmap/AppMapAgentMojo.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

> Task :generatePluginDescriptor
[main] INFO org.apache.maven.cli.MavenCli - Error stacktraces are turned on.
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - Scanning for projects...
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - 
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - ------------------< com.appland:appmap-maven-plugin >-------------------
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - Building AppMap Maven Plugin 1.1.2
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - ----------------------------[ maven-plugin ]----------------------------
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - 
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - --- maven-plugin-plugin:3.6.0:descriptor (default-cli) @ appmap-maven-plugin ---
[main] INFO org.apache.maven.plugin.plugin.DescriptorGeneratorMojo - Using 'UTF-8' encoding to read mojo source files.
[main] INFO org.apache.maven.tools.plugin.scanner.DefaultMojoScanner - java-javadoc mojo extractor found 0 mojo descriptor.
[main] INFO org.apache.maven.tools.plugin.scanner.DefaultMojoScanner - java-annotations mojo extractor found 2 mojo descriptors.
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - ------------------------------------------------------------------------
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - BUILD SUCCESS
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - ------------------------------------------------------------------------
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - Total time:  1.046 s
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - Finished at: 2021-04-28T17:26:01-04:00
[main] INFO org.apache.maven.cli.event.ExecutionEventLogger - ------------------------------------------------------------------------

> Task :compileTestJava
warning: unknown enum constant LifecyclePhase.TEST_COMPILE
  reason: class file for org.apache.maven.plugins.annotations.LifecyclePhase not found
warning: unknown enum constant LifecyclePhase.TEST_COMPILE
2 warnings

> Task :test FAILED

com.appland.appmap.PluginLoadTest > testMojoCanBeInstantiated FAILED
    java.io.FileNotFoundException at PluginLoadTest.java:31

com.appland.appmap.PluginLoadTest > testMojoFailsWithMissingDefaultConfigFile FAILED
    java.io.FileNotFoundException at PluginLoadTest.java:53

com.appland.appmap.PluginLoadTest > testMojoIsLoaded FAILED
    java.io.FileNotFoundException at PluginLoadTest.java:42

com.appland.appmap.PluginLoadTest > testMojoFailsWithNotExistingConfigFile FAILED
    java.io.FileNotFoundException at PluginLoadTest.java:69

com.appland.appmap.GoalLoadTest > testValidateConfigGoalIsLoadedCorrectly FAILED
    java.io.FileNotFoundException at GoalLoadTest.java:21

com.appland.appmap.GoalLoadTest > testPrepareAgentGoalIsLoadedCorrectly FAILED
    java.io.FileNotFoundException at GoalLoadTest.java:14

7 tests completed, 6 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///Users/dustin/dev/appland/appmap-maven-plugin/build/reports/tests/test/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.6.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 4s
10 actionable tasks: 10 executed
```

**Comment by marcuss** [comment]
@dustinbyrne  Finally saw the issue happening and corrected it is a bit cumbersome but this line from pom.xml

https://github.com/applandinc/appmap-maven-plugin/pull/16/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L7

must be in synch with the gradle. build version of the project.

Shouldn't we use maven to build a maven plugin?

In any case, we can move forward with this now.

**Comment by marcuss** [comment]
@dustinbyrne To further explain this, there is a moment in the GoalLoadTest that a maven project is built from a test pom file in the test resources folder that actually uses the maven plugin as defined in the root maven pom.xml file, so altho the project CD/CI pipeline uses gradle.build file to built, the pom.xml will still need to be kept and kept updated... shouldn't we just use it to build the project?

**Comment by marcuss** [comment]
@dustinbyrne during the weekend I added code for:
- Fix issue 13
- Additional tests for issue 5
- Applied the google formatting style to the code.

**Pull request closed without merging**